main.go

```diff
@@ -80,212 +80,243 @@ func main() {
 		case <-done:
-			// wait for all handlers to terminate
+			// wait for all updates are processed
 			wg.Wait()
 			return
 		}
 	}
 }
 
-// handleService handles service updates in a consistent way.
-// It will run until service is deleted or done is closed.
-func handleService(
-	c *configuration,
+// Handles service updates in a consistent way.
+// It's up until service is deleted or done is closed.
+func handleServiceUpdates(
+	urlMapName string,
 	client cloud.Cloud,
-	tag string,
+	tagConfig *tagConfiguration,
 	updates <-chan *registry.ServiceUpdate,
 	wg *sync.WaitGroup,
 	done chan struct{},
 ) {
-	tagInfo, err := parseTag(tag)
-	if err != nil {
-		glog.Error(err)
-		return
-	}
-
-	serviceGroupName := tagInfo.String()
+	tagName := tagConfig.Name
 
+	// todo(max): Why do we need lock if all service updates are processed one by one? It makes more sense to have one lock for all services if we update common resource as e.g. URL map.
 	lock := &sync.RWMutex{}
 	isRunning := false
 	instances := make(map[string]*registry.ServiceInstance)
 
+	errorTemplate := "Failed handling of %s update for service with tag %q: %+v"
+
 	for {
 		select {
 		case update := <-updates:
 			switch update.UpdateType {
 			case registry.NEW:
-				lock.Lock()
-				if !isRunning {
-					glog.Infof("Initializing service with tag %s ...", tag)
-
-					negName := makeName("neg", serviceGroupName)
-					glog.Infof("Creating network endpoint group %s ...", negName)
-					if err := client.CreateNetworkEndpointGroup(negName); err != nil {
-						glog.Errorf("Can't create network endpoint group %s: %v", negName, err)
-						lock.Unlock()
-						continue
-					}
-
-					tagConfig, err := c.GetTagConfiguration(tag)
-					if err != nil {
-						glog.Error(err)
-						continue
-					}
-					hcName := makeName("hc", serviceGroupName)
-					glog.Infof("Creating health check %s ...", hcName)
-					if err := client.CreateHealthCheck(hcName, tagConfig.HealthCheck.Path); err != nil {
-						glog.Errorf("Can't create health check %s: %v", hcName, err)
-						lock.Unlock()
-						continue
-					}
-
-					bsName := makeName("bs", serviceGroupName)
-					glog.Infof("Creating backend service %s ...", bsName)
-					if err = client.CreateBackendService(
-						bsName,
-						negName,
-						hcName,
-						tagInfo.Affinity,
-						tagInfo.CDN,
-					); err != nil {
-						glog.Errorf("Can't create backend service %s: %v", bsName, err)
-						lock.Unlock()
-						continue
-					}
-
-					glog.Infof("Updating URL map %s ...", c.Cloud.URLMap)
-					if err := client.UpdateURLMap(
-						c.Cloud.URLMap,
-						bsName,
-						tagInfo.Host,
-						tagInfo.Path,
-					); err != nil {
-						glog.Errorf("Can't update URL map %s: %v", c.Cloud.URLMap, err)
-						lock.Unlock()
-						continue
-					}
-
-					isRunning = true
-					glog.Infof("Watching service with tag [%s]", tag)
+				if err := handleServiceNewUpdate(urlMapName, client, tagConfig, lock, &isRunning); err != nil {
```

Comment:
Please don't pass boolean value by pointer in the mtehod, it creates a very cryptic logic, as it really hard to understand where this value is changed.